### ISSUE TITLE
MyqlException (0x80004005) - Too many connections

### ISSUE THREAD

**Issue by kryptobi** [PomeloFoundation/Pomelo.EntityFrameworkCore.MySql#1778]
we have an efcore application with the pomelo database driver.
After some time we keep getting the message "Too many connections"

We can't explain exactly how this situation comes about?

We didn't specify pooling explicitly. The default of 100 would apply here. However, if we look at the connections via our database monitor, we have a maximum of approx. 61 connections.
This makes the search more difficult because 61 < 100.

```
Exception message:
Stack trace:
Grant validation error: Too many connections\nMySqlConnector.MySqlException (0x80004005): Too many connections\n at MySqlConnector.Core.ServerSession.ConnectAsync(ConnectionSettings cs, MySqlConnection connection, Int32 startTickCount, ILoadBalancer loadBalancer, Activity activity, IOBehavior ioBehavior, CancellationToken cancellationToken) in //src/MySqlConnector/Core/ServerSession.cs:line 604\n at 
MySqlConnector.Core.ConnectionPool.ConnectSessionAsync(MySqlConnection connection, String logMessage, Int32 startTickCount, Activity activity, IOBehavior ioBehavior, CancellationToken cancellationToken) in //src/MySqlConnector/Core/ConnectionPool.cs:line 468\n at 
MySqlConnector.Core.ConnectionPool.ConnectSessionAsync(MySqlConnection connection, String logMessage, Int32 startTickCount, Activity activity, IOBehavior ioBehavior, CancellationToken cancellationToken) in //src/MySqlConnector/Core/ConnectionPool.cs:line 468\n at 
MySqlConnector.Core.ConnectionPool.GetSessionAsync(MySqlConnection connection, Int32 startTickCount, Activity activity, IOBehavior ioBehavior, CancellationToken cancellationToken) in //src/MySqlConnector/Core/ConnectionPool.cs:line 130\n at MySqlConnector.Core.ConnectionPool.GetSessionAsync(MySqlConnection connection, Int32 startTickCount, Activity activity, IOBehavior ioBehavior, CancellationToken cancellationToken) in //src/MySqlConnector/Core/ConnectionPool.cs:line 130\n at MySqlConnector.MySqlConnection.CreateSessionAsync(ConnectionPool pool, Int32 startTickCount, Activity activity, Nullable1 ioBehavior, CancellationToken cancellationToken) in /_/src/MySqlConnector/MySqlConnection.cs:line 966\n   at MySqlConnector.MySqlConnection.OpenAsync(Nullable1 ioBehavior, CancellationToken cancellationToken) in //src/MySqlConnector/MySqlConnection.cs:line 457\n at MySqlConnector.MySqlConnection.Open() in //src/MySqlConnector/MySqlConnection.cs:line 382\n at Microsoft.EntityFrameworkCore.ServerVersion.AutoDetect(String 
```

### Further technical details

MySQL version: 5.5.5 - 10.6.12
Pomelo.EntityFrameworkCore.MySql version: 7.0.0
Microsoft.AspNetCore.App version: net7

**Comment by yukozh** [comment]
@bgrainger

**Comment by kryptobi** [comment]
Yes but with no effects

**Comment by bgrainger** [comment]
What if you set it to 10? Do you still log approx 61 connections on the server in that case?

**Comment by yukozh** [comment]
> What if you set it to 10? Do you still log approx 61 connections on the server in that case?

The usage will never exceeds the capacity. If set it to 200, I got approx 160 connections. But the exception 'Too many connections' exception will be thrown.

**Comment by bgrainger** [comment]
The server is closing the connection with an error; MySqlConnector is simply reporting this fatal error back to you.

> I think this should not be a user-error.

What should it be instead?

You could try using [Polly](https://github.com/App-vNext/Polly) to implement a retry policy around opening a connection.

**Comment by kryptobi** [comment]
Could the polly solution from this issue meet the retry policy?

https://github.com/PomeloFoundation/Pomelo.EntityFrameworkCore.MySql/issues/1368

**Comment by bgrainger** [comment]
> Took a smaller one, and sent it to your email.

I just found this in my Spam folder; will take a look.

**Comment by bgrainger** [comment]
> Even set Pooling=false, still not work, the error Too many connections is still happen. 

This will cause the problem. When Pooling is disabled, nothing will enforce a limit on concurrent connections. Each `MySqlConnection` you open will make its own connection to the backend server, which could easily exceed the maximum number of allowed server connections.

**Comment by yukozh** [comment]
I'll check the connection leaks in pomelo

**Comment by kryptobi** [comment]
@lauxjpn @yukozh 
I think that I have found the error. you are right. Its the user code.

We use the quartz scheduler (http://www.quartz-scheduler.org/) for our cron jobs. After debugging, we noticed that the instances are never disposed because the quartz schedular uses the root scope.

Within the scheduler we have opened several connections to a third database, although it would be correct to first check whether there is already a connection.

This means that the quartz scheduler did not dispose its instances including connections correctly and we had too many new connections due to the instances.

After we solved the registration of the quartz instances via singletons and checked whether there was still an open connection to the third database, we no longer received the error message.

But i will try your tutorial to check how many connections we got.
